modules/core/src/main/java/org/apache/ignite/internal/util/nio/compression/GridNioCompressionFilter.java

```diff
@@ -0,0 +1,330 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.util.nio.compression;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import javax.cache.configuration.Factory;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.util.nio.GridNioException;
+import org.apache.ignite.internal.util.nio.GridNioFilterAdapter;
+import org.apache.ignite.internal.util.nio.GridNioFinishedFuture;
+import org.apache.ignite.internal.util.nio.GridNioFuture;
+import org.apache.ignite.internal.util.nio.GridNioSession;
+import org.apache.ignite.internal.util.nio.ssl.GridNioSslFilter;
+import org.apache.ignite.lang.IgniteInClosure;
+
+import static org.apache.ignite.internal.util.nio.GridNioSessionMetaKey.COMPRESSION_META;
+
+/** */
+public final class GridNioCompressionFilter extends GridNioFilterAdapter {
+    /** Logger to use. */
+    private final IgniteLogger log;
+
+    /** Order. */
+    private final ByteOrder order;
+
+    /** Allocate direct buffer or heap buffer. */
+    private final boolean directBuf;
+
+    /** Whether direct mode is used. */
+    private boolean directMode;
+
+    /** */
+    private final Factory<CompressionEngine> compressionFactory;
+
+    /**
+     * Creates compress filter.
+     *
+     * @param compressionFactory Factory.
+     * @param directBuf Direct buffer flag.
+     * @param order Byte order.
+     * @param log Logger to use.
+     */
+    public GridNioCompressionFilter(Factory<CompressionEngine> compressionFactory,
+        boolean directBuf,
+        ByteOrder order,
+        IgniteLogger log) {
+        super("Compression filter");
```

Comment:
Add assertions for `compressionFactory`, `order`, `log`.

Reply:
Done.